### PR TITLE
Allergen CE Conversion

### PR DESCRIPTION
**Pull request by KillianKirilenko**
Consider all numbers to be testing placeholders. Needs some tuning. Tested as a skrell by eating two bites of regular meat: this briefly incapacitated the test dummy and left them functionally crippled for a short duration, but __wasn't__ fatal.

Some long overdue updates and improvements to the allergen system.
- Converts allergen reactions to be based off the chemical effects system; if a reagent is an allergen to the mob, it simply adds the allergen chem_effect to their list of chem_effects, and actual effect processing now occurs in Life() rather than on the reagent's end.
- Adds Brute Damage, Burn Damage, and Confusion to the list of possible effects. Weaken is replaced with Pain and Confusion in the default reactions: no more instant collapse, but you're still vulnerable.
- Adds a remove effect proc so that reagents or procs can remove/reduce existing effects on the target. If a magnitude is specified they'll reduce it by that amount (to a minimum of 0), else they'll just flatten it entirely. Could be used to have sedatives counteract hyperzine (or vice versa) and other interactions.
- Makes inaprovaline counter immediate allergic reactions as they occur, so there's a reliable means of treating reactions besides "carbon and pray". You can either administer it ahead of time if you expect to encounter an allergen, or quickly administrate to an ongoing reaction to halt it. If they're still taking damage, administer inaprovaline: if not, administer dylovene and painkillers.

Treatment of an *ongoing* allergic reaction is now "apply inaprovaline". Treatment of one that *already happened* is now dylo+painkillers. Your window to apply inaprovaline will be fairly small, but if you're quick you can head off a reaction at the pass and at least keep it from getting any worse.

As a bonus if you apply inaprovaline _ahead_ of taking an allergen you can prevent the reaction entirely and enjoy what would otherwise be deadly poison (for as long as the inaprovaline is in your system), though you still won't gain any nutrition or other benefits.

:cl:
tweak - standard allergic reactions now use the chemical effects system, and can be prevented or reduced by administering inaprovaline. doesn't apply to special cases like unathi + sugar/etc.
tweak - allergic reactions no longer apply weaken, instead they now cause pain, blurred vision, and confusion. they can still knockdown/incap through paincrit or softcrit.
/:cl:

### CODE REVIEW

File: code/modules/mob/living/carbon/human/species/species.dm [hunk 1]
```diff
@@ -53,9 +53,9 @@
 
 	var/taste_sensitivity = TASTE_NORMAL							// How sensitive the species is to minute tastes.
 	var/allergens = null									// Things that will make this species very sick
-	var/allergen_reaction = AG_TOX_DMG|AG_OXY_DMG|AG_EMOTE|AG_PAIN|AG_WEAKEN		// What type of reactions will you have? These the 'main' options and are intended to approximate anaphylactic shock at high doses.
-	var/allergen_damage_severity = 1.2							// How bad are reactions to the allergen? Touch with extreme caution.
-	var/allergen_disable_severity = 2.5							// Whilst this determines how long nonlethal effects last and how common emotes are.
+	var/allergen_reaction = AG_TOX_DMG|AG_OXY_DMG|AG_EMOTE|AG_PAIN|AG_BLURRY|AG_CONFUSE	// What type of reactions will you have? These the 'main' options and are intended to approximate anaphylactic shock at high doses.
+	var/allergen_damage_severity = 4							// How bad are reactions to the allergen? Touch with extreme caution.
+	var/allergen_disable_severity = 10							// Whilst this determines how long nonlethal effects last and how common emotes are.
```
[the remark this answers is not in the document]
Comment (KillianKirilenko) [reply code/modules/mob/living/carbon/human/species/species.dm:58]: I basically had to crank them up to get any real kind of feedback in the moment, was the thing. It looks a bit worse than it is, I think, but I guess that depends on the goal.

Remember that these are also species-level values and can be easily adjusted in each species datum. If you want skrell to *just* get a bit of tox and discomfort, then their set can be overriden to just AG_TOX_DMG|AG_PAIN (and maybe AG_EMOTE), and the damage/disable can be adjusted for them specifically.

If you (or anyone else) has some more specific numbers to work with I can slip the changes in with another push.

File: code/modules/mob/living/carbon/carbon.dm [hunk 2]
```diff
@@ -435,6 +435,12 @@
 	else
 		chem_effects[effect] = magnitude
 
+/mob/living/carbon/proc/remove_chemical_effect(var/effect, var/magnitude)
+	if(effect in chem_effects && magnitude)
+		chem_effects[effect] = max(0,chem_effects[effect]-magnitude)
+	else if(effect in chem_effects)
+		chem_effects[effect] = 0
```
Comment (Atermonera) [inline code/modules/mob/living/carbon/carbon.dm:442]:
```suggestion
	if(effect in chem_effects)
		chem_effects[effect] = magnitude ? max(0,chem_effects[effect]-magnitude) : 0
```
Note also that a negative magnitude will instead worsen the effect. Probably fine, I think the magnitude bit would make it weird to try to mix with the proc to add effects.

File: code/modules/mob/living/carbon/human/species/species.dm (hunk 1 above)
[the remark this answers is not in the document]
Comment (Atermonera) [reply code/modules/mob/living/carbon/human/species/species.dm:58]: I would keep the damage severity at 2.5, but leave the increase in effect severity at 10, as a middle ground. The transient effects become more pronounced, but long-term damage remains not-lethal.

**Comment by KillianKirilenko** [comment]
Changes made. Also adjusted the changelog notes for extra clarity.

**Comment by Spookerton** [comment]
Should probably have another look over by prior reviewers @Cerebulon @Atermonera

**Review by Atermonera — APPROVED** [review]

**Pull request merged**